[Sema] -Wformat: warn about C23 %b %B when not in C23 language mode

drodriguez opened the pull request:
The validation of %b and %B was done without taking into account the current language mode, which means that %b and %B could have been used in format strings before they were supported.

Change the code to check for the language mode being C23 or after, modify some tests that use %b and %B to specify C23 language mode and add a new test that checks the different behaviour depending on the current language mode.

drodriguez committed:
[Sema] -Wformat: warn about C23 %b %B when not in C23 language mode

The validation of %b and %B was done without taking into account the
current language mode, which means that %b and %B could have been used
in format strings before they were supported.

Change the code to check for the language mode being C23 or after,
modify some tests that use %b and %B to specify C23 language mode and
add a new test that checks the different behaviour depending on the
current language mode.

drodriguez commented:
/cc @SharonXSharon

Fznamznon commented on clang/test/Sema/format-strings-c23-binary.c:
```diff
@@ -0,0 +1,41 @@
+
+int printf(const char *restrict, ...);
+
+#if PRE_C23
```
Why not use `-verify=c23` and then `c23-warning {{...}}` instead of macros?

Fznamznon replied:
Also, should it be a compat warning in pre-c23 mode instead?

Fznamznon reviewed (commented):
Perhaps also a release note makes sense so the users know about the change.

AaronBallman commented:
> The validation of %b and %B was done without taking into account the current language mode, which means that %b and %B could have been used in format strings before they were supported.

On the one hand, yes, this is correct. On the other hand, a C23-based standard library is not going to know that the host application was compiled in C17 mode and it may very well support these modifiers: https://godbolt.org/z/G7KdeKqEx So the issue is that in C23 mode, we can assume the standard library is C23-compatible, but in pre-C23 modes, we can't make assumptions one way or the other.

So I think we don't want to treat the modifiers as being unknown. Instead, I think we should issue a diagnostic along the lines `format specifier '%b' requires a C standard library that is compatible with C23; data argument may not be used by format`. I think this is a different kind of incompatibility than we've had before, so it may require more invasive changes. However, we'll need this for other format specifiers like `%wN` and `%wfN`.

SharonXSharon commented:
> /cc @SharonXSharon

thanks for submitting the fix! A noob question, I wonder in general why does clang choose to warn on the invalid formatter instead of erroring out given it'll be undefined behavior for invalid formatter?

drodriguez commented:
- Rebased into a current `main`.
- Changed the test to use `-verify=c17` and `-verify=c23` and have only one version of the code and no defines.
- Changed the implementation to warn when C23 is not used and the conversion specifiers might not be supported (instead of being unknown). Hope the place I found for the check is a good one. I also changed the wording a little to match other similar messages (no `%b`, but just `'b'`, which seems what's done in other messages).

AaronBallman reviewed (commented):
Changes generally LGTM but the test failures are related.

drodriguez committed:
Adapt format-strings.c to handle both pre-C23 and post-C23 warnings

drodriguez committed:
Remove unnecessary formatting change

michaelrj-google commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10631,6 +10631,9 @@ def warn_printf_data_arg_not_used : Warning<
   "invalid conversion specifier '%0'">, InGroup<FormatInvalidSpecifier>;
+def warn_format_conversion_specifier_requires_c23 : Warning<
+  "conversion specifier '%0' requires a C standard library compatible with "
+  "C23; data argument may not be used by format">, InGroup<Format>;
```
Currently for LLVM-libc our support for `%b` is decided when you compile the library. If you change the `-std=c23` flag when compiling your user code that won't change the validity of using `%b`. It looks like glibc does have mechanisms in its headers to select different function variants based on the version of C you're compiling for (e.g. for `strtol`) but it doesn't seem to use those for `printf`.

In conclusion: It seems like `%b` support from the libc side is more related to if you have a new enough version of the library than the version of C you're compiling for.
